test/unit/util/cloud-provider.test.js

```diff
@@ -80,8 +98,60 @@ describe('CloudProvider', () => {
             method: 'ack',
             name: 'name2'
         });
-        cloudProvider.onMessage({data: `${msg1}\n${msg2}`});
+        cloudProvider.connection._receive({data: `${msg1}\n${msg2}`});
         expect(vmIOData[0].varUpdate.name).toEqual('name1');
         expect(vmIOData[1].varCreate.name).toEqual('name2');
     });
+
+    test('connecting sets connnection attempts back to 1', () => {
+        expect(cloudProvider.connectionAttempts).toBe(0);
+        cloudProvider.connectionAttempts = 10;
+        cloudProvider.connection._open();
+        expect(cloudProvider.connectionAttempts).toBe(1);
+    });
+
+    test('disconnect waits for a period equal to 2^k-1 before trying again', () => {
+        websocketConstructorCount = 1; // This is global, so set it back to 1 to start
+
+        // Connection attempts should still be 0 because connection hasn't opened yet
+        expect(cloudProvider.connectionAttempts).toBe(0);
+        cloudProvider.connection._open();
+        expect(cloudProvider.connectionAttempts).toBe(1);
+
+        cloudProvider.connection.close();
+        expect(timeout).toEqual(1 * 1000); // 2^1 - 1
+        expect(websocketConstructorCount).toBe(2);
+        expect(cloudProvider.connectionAttempts).toBe(2);
+
+        cloudProvider.connection.close();
+        expect(timeout).toEqual(3 * 1000); // 2^2 - 1
+        expect(websocketConstructorCount).toBe(3);
+        expect(cloudProvider.connectionAttempts).toBe(3);
+
+        cloudProvider.connection.close();
+        expect(timeout).toEqual(7 * 1000); // 2^3 - 1
+        expect(websocketConstructorCount).toBe(4);
+        expect(cloudProvider.connectionAttempts).toBe(4);
+
+        cloudProvider.connection.close();
+        expect(timeout).toEqual(15 * 1000); // 2^4 - 1
+        expect(websocketConstructorCount).toBe(5);
+        expect(cloudProvider.connectionAttempts).toBe(5);
+
+        cloudProvider.connection.close();
+        expect(timeout).toEqual(31 * 1000); // 2^5 - 1
+        expect(websocketConstructorCount).toBe(6);
```

Comment:
Should these be changing if we've maxed out at 5? E.g. why is the web socket constructor being called again after maxing out?

Reply:
The websocket constructor count is just how many times we've created the websocket connection object (i.e. tried to open a socket). So it keeps increasing because we keep trying, the only thing that maxes out is the time difference between attempts (i.e. connectionAttempt)

Reply:
I would think that we should stop trying after a while, but I'm not sure... O.o

Reply:
Never know when it'll come back!